iroh-gossip/src/proto/plumtree.rs

```diff
@@ -0,0 +1,603 @@
+
+            // emit event to application
+            io.push(OutEvent::EmitEvent(Event::Received(message.content)));
+
+            self.stats.max_last_delivery_hop =
+                self.stats.max_last_delivery_hop.max(message.round.0);
+        }
+    }
+
+    fn optimize_tree(
+        &mut self,
+        sender: &PA,
+        message: &Gossip,
+        previous_ihaves: VecDeque<(PA, Round)>,
+        io: &mut impl IO<PA>,
+    ) {
+        let round = message.round;
+        let best_ihave = previous_ihaves
+            .iter()
+            .min_by(|(_froma, ra), (_fromb, rb)| ra.cmp(rb))
+            .copied();
+
+        if let Some((ihave_node, ihave_round)) = best_ihave {
+            if (ihave_round < round) && (round - ihave_round) >= self.config.optimization_threshold
+            {
+                let message = Message::Graft(Graft {
+                    id: None,
+                    round: ihave_round,
+                });
+                io.push(OutEvent::SendMessage(ihave_node, message));
+                io.push(OutEvent::SendMessage(*sender, Message::Prune));
+            }
+        }
+    }
+
+    fn on_prune(&mut self, sender: PA) {
+        self.add_lazy(sender);
+    }
+
+    // "When a node receives a IHAVE message, it simply marks the corresponding message as missing
+    // It then starts a timer, with a predefined timeout value, and waits for the missing message to be
+    // received via eager push before the timer expires. The timeout value is a protocol parameter
+    // that should be configured considering the diameter of the overlay and a target maximum recovery latency, defined
+    // by the application requirements. This is a parameter that should be statically configured at deployment time." (p8)
+    fn on_ihave(&mut self, sender: PA, ihaves: Vec<IHave>, io: &mut impl IO<PA>) {
+        for ihave in ihaves {
+            if !self.received_messages.contains(&ihave.id) {
+                self.missing_messages
+                    .entry(ihave.id)
+                    .or_default()
+                    .push_back((sender, ihave.round));
+
+                if !self.graft_timer_scheduled.contains(&ihave.id) {
+                    self.graft_timer_scheduled.insert(ihave.id);
+                    io.push(OutEvent::ScheduleTimer(
+                        self.config.graft_timeout_1,
+                        Timer::SendGraft(ihave.id),
+                    ));
+                }
+            }
+        }
+    }
+
+    fn on_send_graft_timer(&mut self, id: MessageId, io: &mut impl IO<PA>) {
+        if self.received_messages.contains(&id) {
+            return;
+        }
+        let entry = self
+            .missing_messages
+            .get_mut(&id)
+            .and_then(|entries| entries.pop_front());
+        if let Some((peer, round)) = entry {
+            self.add_eager(peer);
+            let message = Message::Graft(Graft {
+                id: Some(id),
+                round,
+            });
```

Comment:
it would be good to verify that the peer is not already eager to prevent sending extra grafts (in principle we could have received or performed a graft for X/Y reason between the creation of the timer and it's deadline)

Reply:
That's a good question, I'm actually not sure: Let's say we receive IHaves A and B both from peer X, and IHave for B also from peer Y (later). The timer for A is triggered, and we send a Graft to peer X. The timer for B is triggered. We want the message content for B. Should we request (graft) it from peer X or not? In your interpretation, we should not, because by now peer X is already eager. However, if peer X is the better candidate to be grafted (because its IHave was received earlier than that of peer Y, wouldn't it be better to just also request message B from peer X (by sending a Graft) instead of also grafting Y (which was slower to deliver to us than X)?

Reply:
I think I'm forgetting that grafting means not only "upgrade our link" but also "gimme the message". In that sense leaving it as it seems fine to me.

However the discussion gets interesting because if we overwhelm peer X with "gimme" requests, it will also make sense to load balance with Y to get the rest. To be clear I think this can be a further optimization we likely don't want to _eargerly_ get in :drum: